src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/BatchPatternMaskerProviderTest.java

```diff
@@ -0,0 +1,367 @@
+
+public class BatchPatternMaskerProviderTest {
+
+    private static final String SIMPLE = "abcABC123";
+    private static final String SAMPLE_PASSWORD = "}#T14'GAz&H!{$U_";
+    private static final String ESCAPE = "^<^>^(^)^^^&^|";
+
+    // ALL_ASCII - [ < > " ' ^ $ | % ] 
+    private static final String NON_DANGEROUS = "!#$*+,-./ 0123456789:;=? @ABCDEFGHIJKLMNO PQRSTUVWXYZ[\\]_ `abcdefghijklmno pqrstuvwxyz{}~";
+
+    // <>"'^$|
+    private static final String NON_DANGEROUS_IN_DOUBLE = "abc<def>$ghi|jkl";
+
+    // quoted form: "^^|\a\\"
+    private static final String NEEDS_QUOTING = "^|\\a\\";
+
+    private static final String ALL_ASCII = "!\"#$%&'()*+,-./ 0123456789:;<=>? @ABCDEFGHIJKLMNO PQRSTUVWXYZ[\\]^_ `abcdefghijklmno pqrstuvwxyz{|}~";
+
+    @Rule
+    public JenkinsRule j = new JenkinsRule();
+
+    private WorkflowJob project;
+
+    private String credentialPlainText;
+    private String credentialId;
+
+    @Before
+    public void assumeWindowsForBatch() throws Exception {
+        assumeTrue(Functions.isWindows());
+    }
+
+    private void registerCredentials(String password) throws IOException {
+        this.credentialPlainText = password;
+        this.credentialId = CredentialsTestUtil.registerStringCredentials(j.jenkins, password);
+    }
+
+    private WorkflowRun runProject() throws Exception {
+        return j.assertBuildStatusSuccess(project.scheduleBuild2(0));
+    }
+
+    @Test
+    public void simple_noQuote() throws Exception {
+        registerCredentials(SIMPLE);
+        assertDirectNoPlainTextButStars(runDirectNoQuote());
+    }
+
+    @Test
+    public void simple_singleQuote() throws Exception {
+        registerCredentials(SIMPLE);
+        assertDirectNoPlainTextButStars(runDirectSingleQuote());
+    }
+
+    @Test
+    public void simple_doubleQuote() throws Exception {
+        registerCredentials(SIMPLE);
+        assertDirectNoPlainTextButStars(runDirectDoubleQuote());
+    }
+
+    @Test
+    public void simple_delayed() throws Exception {
+        registerCredentials(SIMPLE);
+        assertDelayedNoPlainTextButStars(runDelayedAllQuotes());
+    }
+
+    @Test
+    public void nonDangerous_noQuote() throws Exception {
+        registerCredentials(NON_DANGEROUS);
+        assertDirectNoPlainTextButStars(runDirectNoQuote());
+    }
+
+    @Test
+    public void nonDangerous_singleQuote() throws Exception {
+        registerCredentials(NON_DANGEROUS);
+        assertDirectNoPlainTextButStars(runDirectSingleQuote());
+    }
+
+    @Test
+    public void nonDangerous_doubleQuote() throws Exception {
+        registerCredentials(NON_DANGEROUS);
+        assertDirectNoPlainTextButStars(runDirectDoubleQuote());
+    }
+
+    @Test
+    public void nonDangerous_delayed() throws Exception {
+        registerCredentials(NON_DANGEROUS);
+        assertDelayedNoPlainTextButStars(runDelayedAllQuotes());
+    }
+
+    // we do NOT support dangerous characters in direct expansion mode
```

Comment:
Sorry, what is “direct expansion mode” and what is not supported? `help.html` mentions nothing of the sort.

Reply:
I think this refers to what `setupNoQuoteProject` does.

Reply:
I am still unclear on what “direct expansion mode” means, or what this means:

> these commands would be completely broken if echo didn't work the way it does when CREDENTIALS contains special characters

What is the way that echo does work, or how else would it work, and what commands would be “broken” and how?

Reply:
Direct expansion is the default mode; delayed expansion is the newer cmd feature that defers expansion of environment variables until the command is executed instead of while it's parsed.